### PR TITLE
ARROW-1452: [C++] Make macros in arrow/util/macros.h more unique

### PR DESCRIPTION
**Pull request by wesm**
This fixes a bug encounted in third party use: https://github.com/mapd/mapd-core/blob/master/QueryEngine/ResultSet.h#L37

cc @asuhan

### CODE REVIEW

**Comment by pcmoritz** [comment]
+1 we also ran into this in the past :)

**Comment by wesm** [comment]
This is going to fail until PARQUET-1085 https://github.com/apache/parquet-cpp/pull/386 is in

**Comment by wesm** [comment]
@pcmoritz if you're available today could you look at https://github.com/apache/parquet-cpp/pull/386? since we are RTC in Parquet I need someone to look at that before I can merge it

**Comment by wesm** [comment]
THanks @cpcloud  -- triggered build here, so this will hopefully pass now

**Comment by wesm** [comment]
Appveyor build here: https://ci.appveyor.com/project/wesm/arrow/build/1.0.1009. Will merge once that passes

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -254,9 +254,6 @@ class ARROW_EXPORT Array {
     }
     data_ = data;
   }
-
- private:
```
Comment (xhochy) [inline cpp/src/arrow/array.h]: Why was this removed?

File: cpp/src/arrow/io/interfaces.h [hunk 2]
```diff
@@ -79,9 +79,6 @@ class ARROW_EXPORT FileInterface {
   FileInterface() {}
   FileMode::type mode_;
   void set_mode(FileMode::type mode) { mode_ = mode; }
-
- private:
-  DISALLOW_COPY_AND_ASSIGN(FileInterface);
```
Comment (xhochy) [inline cpp/src/arrow/io/interfaces.h]: Why was this removed?

File: cpp/src/arrow/type.h [hunk 3]
```diff
@@ -157,9 +157,6 @@ class ARROW_EXPORT DataType {
  protected:
   Type::type id_;
   std::vector<std::shared_ptr<Field>> children_;
-
- private:
-  DISALLOW_COPY_AND_ASSIGN(DataType);
```
Comment (xhochy) [inline cpp/src/arrow/type.h]: Why was this removed?

**Review by xhochy — COMMENTED** [review]
Seems like some `DISALLOW_COPY_AND_ASSIGN` were removed. Was this intentional? If so: why?

**Comment by wesm** [comment]
Sorry about that, this was not intentional -- there were odd failures in MSVC where I didn't have access to a machine to debug, but I will later today. I'll restore them in a patch now and we can debug any problems that come up: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/build/1.0.3116/job/reqo0yjegpgtkdge#L2699
